File: searchhandlers/khc_mansearch.pl
#!/usr/bin/perl
#
#  Script for searching man pages. The result is generated as HTML.
#
#  This file is part of KHelpcenter.
#
#  SPDX-FileCopyrightText: 2002 SuSE Linux AG Nuernberg
#  SPDX-FileContributor: Cornelius Schumacher <cschum@suse.de>
#
#  SPDX-License-Identifier: GPL-2.0-or-later

use strict;

use Getopt::Long;

my ( $words, $maxcount, $lang, $method, $help );

GetOptions (
  'maxcount=i' => \$maxcount,
  'words=s' => \$words,
  'lang=s' => \$lang,
  'method=s' => \$method,
  'help' => \$help
);

if ( $help ) {
  print STDERR "Usage: khc_mansearch.pl --maxcount=n --words=<string> " .
    "--lang=<languagecode> --method=and|or\n";
  exit 1;
}

if ( $method ne 'and' and $method ne 'or' ) {
  print STDERR "Unrecognized method: $method.\n";
  exit 1;
}

if ( !$words ) {
  print STDERR "No search words given.\n";
  exit;
}

if ( !$lang or $lang eq 'C' ) {
  $lang = 'en';
}

if ( $words =~ /[^+]$/ ) {
  # If the query was a single word, perhaps it is the literal name of the command (or a prefix)
  if ( !open( PREFIX, "-|", ( 'whatis', '-w', $words . '*' ) ) ) {
    print "Can't open whatis.\n";
    exit 1;
  }

  # ... or a substring
  if ( !open( NAME, "-|", ( 'whatis', '-w', '*' . $words . '*' ) ) ) {
    print "Can't open whatis.\n";
    exit 1;
  }
}

# Build the apropos command line
my @apropos;
push @apropos, 'apropos';
push @apropos, '-L', $lang;
if ( $method eq 'and' ) {
  push @apropos, '--and';
}
push @apropos, split( '\+', $words );

# Perform search
if ( !open( DESCRIPTION, "-|", @apropos ) ) {
  print "Can't open apropos.\n";
  exit 1;
}
my @results;
my %seen;
while( $_ = (<PREFIX> // <NAME> // <DESCRIPTION>) ) {
#  print "RAW:$_";
  chop;
  /^([^\s]+)\s+\((.*)\)\s+-\s+(.*)$/;
  my $page = $1;
  my $section = $2;
  my $description = $3;

  if ( $page && ! $seen{ $page . $section }++ ) { push @results, [ $page, $section, $description ]; }
}
close PREFIX;
close NAME;
close DESCRIPTION;

my $nummatches = @results;

if ( $nummatches > 0 ) {
  print "<ul>\n";

  my $count = 0;
  for my $result ( @results ) {
    my ( $page, $section, $description ) = @$result;
    my $url = "man:" . $page . "(" . $section . ")";
    print "<li><a href=\"$url\">";
    print "$page($section) - $description</a></li>\n";
    if ( ++$count == $maxcount ) { last; }
  }

  print "</ul>\n";
}

1;

File: searchhandlers/khc_mansearch.py
#!/bin/env python3
# Script for searching man pages. The result is generated as HTML.
#
# Converted from Perl code which was:
# SPDX-FileCopyrightText: 2002 SuSE Linux AG Nuernberg
# SPDX-FileContributor: Cornelius Schumacher <cschum@suse.de>
#
# SPDX-License-Identifier: GPL-2.0-or-later

import argparse
import subprocess
import re
import sys

def main():
    parser = argparse.ArgumentParser(description='Search for man pages.')
    parser.add_argument('--maxcount', type=int, help='Maximum number of results to return')
    parser.add_argument('--words', type=str, required=True, help='Search words')
    parser.add_argument('--lang', type=str, default='en', help='Language code')
    parser.add_argument('--method', type=str, choices=['and', 'or'], required=True, help='Search method')
    args = parser.parse_args()

    if args.lang == 'C':
        args.lang = 'en'

    # Store command results only when they succeed
    command_results = ""

    # Strip leading and trailing whitespace from search query
    words = args.words.strip()

    # If the query is a single word, perhaps it is the literal name of the
    # command, or a prefix, or a substring
    if not ' ' in words:
        results_prefix = subprocess.run(['whatis', '-w', words + '*'], capture_output=True, text=True)
        if results_prefix.returncode == 0:
            command_results += results_prefix.stdout
        results_substring = subprocess.run(['whatis', '-w', '*' + words + '*'], capture_output=True, text=True)
        if results_substring.returncode == 0:
            command_results += results_substring.stdout

    # Build the apropos command line
    apropos = ['apropos', '-L', args.lang]
    if args.method == 'and':
        apropos.append('--and')
    apropos.extend(words.split(' '))

    # Perform search
    results_apropos = subprocess.run(apropos, capture_output=True, text=True)
    if results_apropos.returncode != 0:
        command_results += results_apropos.stdout

    # Results from the previous commands are formatted as:
    # assert_perror (3)    - test errnum and abort
    pattern = re.compile(r'^([a-zA-Z0-9_:-]+)\s+\((.*)\)\s+-\s+(.*)$')

    results = []
    seen = set()
    for line in command_results.splitlines():
        match = pattern.match(line)
        if match:
            page, section, description = match.groups()
            if page and (page + section) not in seen:
                seen.add(page + section)
                results.append((page, section, description))

    if results:
        print("<ul>")
        count = 0
        for page, section, description in results:
            url = f"man:{page}({section})"
            print(f"<li><a href=\"{url}\">{page}({section}) - {description}</a></li>")
            count += 1
            if count == args.maxcount:
                break
        print("</ul>")

if __name__ == "__main__":
    main()
